src/server.ts

```diff
@@ -1,13 +1,67 @@
-import express from "express";
+import express from 'express';
+import { readFileSync } from 'fs';
+import axios, {AxiosResponse} from 'axios';
+
+class MarsApp {
+    
+    app = express();
+    port = 8000;
+    router = express.Router();
+
+    // path to the NASA API key
+    KEY_PATH = 'res/key';
+
+    key : string = "";
+    
+    ReadAPIKey(): string {
+        let ret : string = readFileSync(this.KEY_PATH).toString();
+        return ret;
+    }
+
+    /*
+     * Returns the list of rovers in JSON format.
+     * Sends a GET request to api.nasa.gov.
+     * Requires an API key.
+     */
+    async GetRoverList(key: string) : Promise<string> {
+        const roverPath = 'https://api.nasa.gov/mars-photos/api/v1/rovers' + '?api_key=' + key;
+        let ret : string = "";
+
+        try {
+            const response = await axios.get(roverPath);
+            ret = response.data;
+        } catch (exception) {
+            console.error(exception);
+        }
+
+        return ret;
```

Comment:
If something does actually go wrong and an exception is thrown, what should this function return, _if anything_?

At the moment, it returns an empty string, which implies that there are simply no rovers!

Reply:
Yes, I think I overcomplicated that code. I rewrote it in a simpler manner now.